### PR TITLE
[red-knot] Handle invalid assignment targets

### LINKED ISSUE #14321: [red-knot] Panics on named expressions with invalid targets
Red knot currently panics on three out of four of these examples from `crates/ruff_python_parser/resources/invalid/expressions/named/invalid_target.py`, at two different places:
```py
(x.y := 1)  # panicked at crates/red_knot_python_semantic/src/types/infer.rs:2521:37:
(x[y] := 1)  # panicked at crates/red_knot_python_semantic/src/types/infer.rs:2521:37:
(*x := 1)
([x, y] := [1, 2])  # panicked at crates/red_knot_python_semantic/src/semantic_index/builder.rs:224:9:
```
We currently don't handle `ExprNamed::target`s that are not an `ExprName` properly.

### LINKED ISSUE #14322: [red-knot] Panics on invalid annotated tuple assignment
Red knot currently panics if someone tries to do this:
```py
(x, y): tuple[int, int] = (2, 3)
```
The same panic appears in this seemingly different example … 
```py
def f^(x, y):
    pass
```
where an invalid character triggers error recovery in the parser, and the remaining `(x, y):` part appears to be an annotated assignment to a tuple.

The assertion is caused by the fact that we call `SemanticIndexBuilder::add_definition` twice, once for each tuple element.

### PR DESCRIPTION
**Pull request by sharkdp**
## Summary

This fixes several panics related to invalid assignment targets. All of these led to some a crash, previously:
```py
(x.y := 1)  # only name-expressions are valid targets of named expressions
([x, y] := [1, 2])  # same
(x, y): tuple[int, int] = (2, 3)  # tuples are not valid targets for annotated assignments
(x, y) += 2  # tuples are not valid targets for augmented assignments
```

closes #14321
closes #14322

## Test Plan

I symlinked four files from `crates/ruff_python_parser/resources/{invalid,inline/err}` into the red knot corpus, as they seemed like ideal test files for this exact scenario. I can also copy them, if we prefer that (these relative symlinks are probably annoying if we move folders around). I think eventually, it might be a good idea to simply include *all* invalid-syntax examples from the parser tests into red knots corpus (I believe we're actually not too far from that goal). Or expand the scope of the corpus test to this directory. Then we can get rid of these symlinks again.

### CODE REVIEW

File: crates/red_knot_python_semantic/src/semantic_index/builder.rs [hunk 1]
```diff
@@ -690,9 +699,21 @@ where
...
-                self.visit_expr(target);
-                self.pop_assignment();
+
+                // See https://docs.python.org/3/library/ast.html#ast.AugAssign
+                if matches!(
+                    **target,
+                    ast::Expr::Attribute(_)
+                        | ast::Expr::Subscript(_)
+                        | ast::Expr::Starred(_)
+                        | ast::Expr::Name(_)
```
Comment (sharkdp) [inline crates/red_knot_python_semantic/src/semantic_index/builder.rs:709]: I tried to search for something like `Expr::is_valid_{aug_,ann_,}assign_target` methods but couldn't find anything. And all existing code that deals with assignments could not make use of such a function, as it typically needs to do something specific for each branch. So it's just an inline match for now.

File: crates/red_knot_workspace/resources/test/corpus/04_assign_invalid_targets.py
Comment (AlexWaygood) [inline crates/red_knot_workspace/resources/test/corpus/04_assign_invalid_targets.py:1]: is this the test file you wanted to add?

File: crates/red_knot_python_semantic/src/semantic_index/builder.rs [hunk 2]
```diff
@@ -690,9 +699,21 @@ where
...
-                self.push_assignment(aug_assign.into());
-                self.visit_expr(target);
-                self.pop_assignment();
+
+                // See https://docs.python.org/3/library/ast.html#ast.AugAssign
+                if matches!(
+                    **target,
+                    ast::Expr::Attribute(_)
+                        | ast::Expr::Subscript(_)
+                        | ast::Expr::Starred(_)
```
Comment (AlexWaygood) [inline crates/red_knot_python_semantic/src/semantic_index/builder.rs:708]: I don't think `Starred` expressions are legal in `AugAssign`s:

The parser permits them in `Assign`s:

```pycon
>>> import ast
>>> print(ast.dump(ast.parse('*y = z'), indent=2))
Module(
  body=[
    Assign(
      targets=[
        Starred(
          value=Name(id='y', ctx=Store()),
          ctx=Store())],
      value=Name(id='z', ctx=Load()))])
```

But not `AugAssign`s:

```pycon
>>> import ast
>>> print(ast.dump(ast.parse('*y += z'), indent=2))
Traceback (most recent call last):
  File "<python-input-4>", line 1, in <module>
    print(ast.dump(ast.parse('*y += z'), indent=2))
                   ~~~~~~~~~^^^^^^^^^^^
  File "/Users/alexw/.pyenv/versions/3.13.0/lib/python3.13/ast.py", line 54, in parse
    return compile(source, filename, mode, flags,
                   _feature_version=feature_version, optimize=optimize)
  File "<unknown>", line 1
    *y += z
    ^^
SyntaxError: 'starred' is an illegal expression for augmented assignment
```

File: crates/red_knot_workspace/resources/test/corpus/04_assign_invalid_targets.py
Comment (sharkdp) [reply crates/red_knot_workspace/resources/test/corpus/04_assign_invalid_targets.py:1]: Yes. I wanted to make sure that invalid *normal* assignments (non-annotated, non-augmented) were also handled correctly. Or what is wrong?

File: crates/red_knot_python_semantic/src/semantic_index/builder.rs (hunk 2 above)
Comment (sharkdp) [reply crates/red_knot_python_semantic/src/semantic_index/builder.rs:708]: Ah, thanks. This means the Python docs are a bit misleading here. For `AugAssign`, they state that the *"target attribute cannot be of class [Tuple](https://docs.python.org/3/library/ast.html#ast.Tuple) or [List](https://docs.python.org/3/library/ast.html#ast.List), unlike the targets of [Assign](https://docs.python.org/3/library/ast.html#ast.Assign)."* — which I interpreted as: everything allowed for a normal assignment, but not `Tuple` or `List`.

File: crates/red_knot_workspace/resources/test/corpus/04_assign_invalid_targets.py
Comment (AlexWaygood) [reply crates/red_knot_workspace/resources/test/corpus/04_assign_invalid_targets.py:1]: It looks from the GitHub diff as though this file contains the literal text

```
../../../../ruff_python_parser/resources/invalid/statements/invalid_assignment_targets.py
```

Am I missing something -- is it a symlink or something? 😄

Comment (sharkdp) [reply crates/red_knot_workspace/resources/test/corpus/04_assign_invalid_targets.py:1]: Yes. See PR description :smile:.

Comment (sharkdp) [reply crates/red_knot_workspace/resources/test/corpus/04_assign_invalid_targets.py:1]: I'm waiting for the CI failure on Windows.

Comment (sharkdp) [reply crates/red_knot_workspace/resources/test/corpus/04_assign_invalid_targets.py:1]: But my links are a bit off. Give me a minute.

File: crates/red_knot_python_semantic/src/semantic_index/builder.rs (hunk 2 above)
Comment (AlexWaygood) [reply crates/red_knot_python_semantic/src/semantic_index/builder.rs]: I agree the docs are bad here.

I think arguably the docs for the stdlib `ast` module say too _much_ here, though. They should instead say less, and link to the canonical docs for which expressions are allowed as augmented-assignment targets, which are at https://docs.python.org/3/reference/simple_stmts.html#augmented-assignment-statements (and which are much more precise)

File: crates/red_knot_workspace/resources/test/corpus/04_assign_invalid_targets.py
Comment (AlexWaygood) [reply crates/red_knot_workspace/resources/test/corpus/04_assign_invalid_targets.py:1]: > Yes. See PR description 😄.

Ahh sorry!!

> But my links are a bit off. Give me a minute.

Yeah, I think GitHub displays correct symlinks differently 😄

Comment (sharkdp) [reply crates/red_knot_workspace/resources/test/corpus/04_assign_invalid_targets.py:1]: > Yeah, I think GitHub displays correct symlinks differently 😄

No I believe the links are correct on a.. filesystem level. But there is a small semantic/naming mistake. This is how GitHub shows symlinks.

Comment (sharkdp) [reply crates/red_knot_workspace/resources/test/corpus/04_assign_invalid_targets.py:1]: Okay. Now it should be fine. The names are different from the source files, but hopefully match the naming pattern of red knot's corpus.

**Comment by carljm** [comment]
I would support just modifying the corpus test to run on all parser examples, whenever we are at a point where that doesn't panic. If we aren't at that point yet, symlinking certain files seems like a fine temporary solution (if it works cross-platform)

**Review by carljm — APPROVED** [review]
Looks great, thank you!

**Pull request merged**